### PR TITLE
Add flash attention to Transformers

### PR DESCRIPTION
**Pull request by Warvito**
Implements #339

### CODE REVIEW

File: generative/networks/blocks/selfattention.py [hunk 1]
```diff
@@ -11,12 +11,22 @@
...
 from __future__ import annotations
 
+import importlib.util
 import math
 
 import torch
 import torch.nn as nn
 from torch.nn import functional as F
 
+if importlib.util.find_spec("xformers") is not None:
```
Comment (marksgraham) [inline generative/networks/blocks/selfattention.py:21]: Currently if the code does not find `xformers` but`use_flash_attention` is set `True` the code errors out. I think we need to self `use_flash_attention=False` in the init if `has_xformers=False`, and ideally raise a warning, too.

File: generative/networks/blocks/selfattention.py [hunk 2]
```diff
@@ -95,15 +108,25 @@ def forward(self, x: torch.Tensor, context: torch.Tensor | None = None) -> torch
...
+            y = xops.memory_efficient_attention(
+                query, key, value, attn_bias=xops.LowerTriangularMask() if self.causal else None
+            )
+
+        else:
+            # manual implementation of attention
+            attention_scores = (query @ key.transpose(-2, -1)) * self.scale
+
+            if self.causal:
+                attention_scores = attention_scores.masked_fill(self.causal_mask[:, :, :t, :kv_t] == 0, float("-inf"))
```
Comment (marksgraham) [inline generative/networks/blocks/selfattention.py:124]: These two methods give different values. `scale` isn't currently passed to `memory_efficient_attention`, nor the dropout probability, but even account for that it looks like that isn't the root of the difference. The python implementation of `memory_efficient_attnetion` looks a little different to the manual implementation here:

https://github.com/facebookresearch/xformers/blob/658ebab39545f180a6075385b3897921623d6c3b/xformers/ops/fmha/__init__.py#L142-L149

**Review by marksgraham — CHANGES_REQUESTED** [review]
Hi Walter,

I've noticed some discrepancies between the two implementations. It would also be good to add tests for `use_flash_attention`, and maybe even a test that compares the output for the calculations with and without flash attention.

File: generative/networks/blocks/selfattention.py (hunk 1 above)
Comment (Warvito) [reply generative/networks/blocks/selfattention.py:21]: Added error message in case user want to use flash attention but xformers is not installed

File: generative/networks/blocks/selfattention.py (hunk 2 above)
Comment (Warvito) [reply generative/networks/blocks/selfattention.py]: thanks for pointing this, I fixed the problem

```
import torch

from generative.networks.blocks import SABlock

device = torch.device("cuda")
sab = SABlock(hidden_size=4, num_heads=2, dropout_rate=0.0, use_flash_attention=False)
sab = sab.to(device)
sab.eval()

with torch.no_grad():
    x = torch.randn(1, 3, 4).to(device)
    result = sab(x)

    sab.use_flash_attention = True
    result_flash = sab(x)

torch.isclose(result, result_flash)
```
returning
```
tensor([[[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]]], device='cuda:0')
```

**Comment by Warvito** [comment]
@marksgraham I was thinking, since now Pytorch 2.0 has native flash attention, do you think it is worth to abandon the `xformer` dependency and adopt the native option instead? Since Pytorch 2.0 has compatibility, it would not be an issue to update to it in place of installing xformers (which requires Pytorch > 1.13)
